Admin menu: Sanitize URLs

mmtr opened the pull request:
#### Changes proposed in this Pull Request

Sanitize the URLs of the admin menu items before rendering to ensure that only internal links are used.

Further reading: p9o2xV-1fe-p2

#### Testing instructions

- Add the code below to the `0-sandbox.php` file in your WP.com sandbox:
```
add_filter( 'rest_post_dispatch', function ( $response ) {
	if ( '/wpcom/v2/sites/(?P<wpcom_site>[\w.:-]+)/admin-menu' === $response->get_matched_route() ) {
		$data = $response->get_data();
		array_unshift( $data, [
			'icon'     => 'dashicons-warning',
			'slug'     => 'my-custom-menu',
			'title'    => 'Click here',
			'type'     => 'menu-item',
			'url'      => 'javascript:alert("hello")',
			'children' => [ [
				'parent' => 'my-custom-menu',
				'slug'   => 'my-custom-menu-2',
				'title'  => 'Or here',
				'type'   => 'submenu-item',
				'url'    => 'http://example.com',
			] ]
		] );
		$response->set_data( $data );
	}
	return $response;
} );
```
- Sandbox the API.
- Go to wordpress.com and note how:
  - Clicking on "Click here" (first menu item) triggers a javascript alert.
  - Clicking on "Or here" (submenu of first menu item) redirects to an external domain.
- Open the Calypso live link from below.
- Make sure that clicking on "Click here" doesn't trigger any javascript alert.
- Make sure that clicking on "Or here" doesn't redirect to an external domain.
- Make sure the rest of the menu items still work as usual. Test with sites with `*.wordpress.com` domains and custom domains.

mmtr committed:
Admin menu: Sanitize URLs

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,38 @@ export const requestFetchAdminMenu = ( action ) =>
-export const handleSuccess = ( { siteId }, menuData ) => {
-	return receiveAdminMenu( siteId, menuData );
+const sanitizeUrl = ( url, site ) => {
+	if ( new RegExp( `^https?://${ site?.domain }` ).test( url ) ) {
```
It might be worth assigning this expression to a `const` with a descriptive name for future devs who may come across this code.

mmtr replied:
Totally, amended in f851171

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,38 @@ export const requestFetchAdminMenu = ( action ) =>
+const sanitizeUrl = ( url, site ) => {
+	if ( new RegExp( `^https?://${ site?.domain }` ).test( url ) ) {
+		return url;
+	} else if ( /^\//.test( url ) ) {
```
It might be worth assigning this expression to a const with a descriptive name for future devs who may come across this code.

mmtr replied:
Totally, amended in f851171

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,38 @@ export const requestFetchAdminMenu = ( action ) =>
+
+const sanitizeMenuItem = ( menuItem, site ) => {
+	let sanitizedChildren;
+	if ( menuItem.children ) {
```
Is it worth checking also that `children` is an array by using `Array.isArray(menuItem.children)` too?

mmtr replied:
Yup, better safe than sorry in this case! Done in ad1c6a2

tjcafferkey commented:
Just made some inline comments. Also noticed that when I clicked the "Click here" menu item I got redirected to the fieldguide with all the menus expanded? Is this expected? Please see screenshot below

![Screenshot 2021-02-05 at 11 50 22](https://user-images.githubusercontent.com/8639742/107031148-9f9e0780-67a9-11eb-8e48-b2e743bccd3f.png)

mmtr committed:
Use descriptive variables for regex checks

mmtr committed:
Check menu items is an array before using .map

mmtr committed:
Handle menu items without URL

mmtr commented:
> Also noticed that when I clicked the "Click here" menu item I got redirected to the fieldguide with all the menus expanded? Is this expected?

Good catch! Turned out to be an issue in how the unified menu handle items without a link. Should've been fixed in 9bfea08

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,43 @@ export const requestFetchAdminMenu = ( action ) =>
+
+const sanitizeMenuItem = ( menuItem, site ) => {
+	let sanitizedChildren;
+	if ( Array.isArray( menuItem.children ) ) {
```
Sorry this is my fault for not being clear.

If `menuItem` is `undefined` this will error perhaps we can optional chain, and do `Array.isArray( menuItem?.children )`

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,43 @@ export const requestFetchAdminMenu = ( action ) =>
+	}
+	return {
+		...menuItem,
+		url: sanitizeUrl( menuItem.url, site ),
```
Shall we put a `?.` check in here `sanitizeUrl( menuItem?.url, site )`

tjcafferkey commented on client/my-sites/sidebar-unified/menu.jsx:
```diff
@@ -47,7 +47,8 @@ export const MySitesSidebarUnifiedMenu = ( {
 
 	const selectedMenuItem =
-		children && children.find( ( menuItem ) => itemLinkMatches( menuItem.url, path ) );
+		children &&
```
Worth doing an array check here as well since we're using `.find`, do you think? `Array.isArray(children) &&`

tjcafferkey commented on client/my-sites/sidebar-unified/menu.jsx:
```diff
@@ -47,7 +47,8 @@ export const MySitesSidebarUnifiedMenu = ( {
 	const selectedMenuItem =
-		children && children.find( ( menuItem ) => itemLinkMatches( menuItem.url, path ) );
+		children &&
+		children.find( ( menuItem ) => menuItem.url && itemLinkMatches( menuItem.url, path ) );
```
Do you think we should check menuItem here? `menuItem?.url`

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,43 @@ export const requestFetchAdminMenu = ( action ) =>
+	return '';
+};
+
+const sanitizeMenuItem = ( menuItem, site ) => {
```
Shall we write some unit tests for this, or create a follow up issue to do so?

tjcafferkey commented on client/state/data-layer/wpcom/sites/admin-menu/index.js:
```diff
@@ -17,8 +18,43 @@ export const requestFetchAdminMenu = ( action ) =>
 
-export const handleSuccess = ( { siteId }, menuData ) => {
-	return receiveAdminMenu( siteId, menuData );
+const sanitizeUrl = ( url, site ) => {
```
Shall we write some unit tests for this, or create a follow up issue to do so?

tjcafferkey commented:
>Good catch! Turned out to be an issue in how the unified menu handle items without a link. Should've been fixed in

I can confirm this now works 😄 thanks!

Just a few other things (sorry!!):
* I've left some inline comments, let me know if you think any of them are worth implementing?
* The unit tests appear to be failing on tests related to menu items. Can we confirm if these are due to these changes or not?

mmtr committed:
Add more safety checks

mmtr committed:
Add tests covering sanitization and fix broken tests

mmtr commented:
All valid feedback, @tjcafferkey! The failing tests were indeed caused by my changes, so I fixed them and added one more covering the new sanitization functions.

mmtr committed:
Collapse rest of menus when link is opened

tjcafferkey reviewed (approved):
Thanks so much for addressing all of that feedback and implementing unit tests @mmtr! All looks good to me and works as described in the testing instructions. 😄 🚀

mmtr commented:
Merging since failing E2E tests are apparently unrelated: p1612456065018600-slack-C1A1EKDGQ

obenland commented:
I wonder if this has broken link URLs for mapped domains in Calypso?

Post > Tags, Post > Categories are empty now for me. Settings > Media, Settings > Reading, too. Probably every wp-admin URL. Is that something you can reproduce?

mmtr commented:
@obenland indeed, it's related to these changes. I would swear I tested this with custom domains and worked fine, so not sure what I missed. I'll fix this today.

mmtr commented:
> @obenland indeed, it's related to these changes. I would swear I tested this with custom domains and worked fine, so not sure what I missed. I'll fix this today.

Turns out that Atomic sites with a mapped domain use the custom domain for their WP Admin URLs (`<CUSTOM_DOMAIN/wp-admin>` while Simple sites with a mapped domain don't use it at all (`<SITE_SLUG>.wordpress.com/wp-admin`).

I wrongly assumed both were using the mapped domain, so the check I introduced was specifically checking the custom domain was part of the URL. I missed this because I only tested the mapped domain scenario with an Atomic sites.

https://github.com/Automattic/wp-calypso/pull/49924 should fix that.
